### PR TITLE
Implement QueryHelper

### PR DESCRIPTION
**Pull request by OlehYakovenko**

### CODE REVIEW

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package com.bobocode.bibernate;
+
+import com.bobocode.bibernate.exception.QueryHelperException;
+import com.bobocode.bibernate.session.Session;
+import com.bobocode.bibernate.session.SessionFactory;
+
+import java.util.function.Function;
+
+public class QueryHelper {
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java:13]: Could you also add a method to persist data?

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    public <T> T readWithinTx(Function<Session, T> entityManagerConsumer) {
+        Session session = sessionFactory.createSession();
+        session.beginTransaction();
+        try {
+            T result = entityManagerConsumer.apply(session);
+            session.commitTransaction();
+            return result;
+        } catch (Exception e) {
+            session.rollbackTransaction();
+            throw new QueryHelperException("Transaction is rolled back", e);
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: In my opinion, we don't need a new custom exception for this class, we can just reuse the generic BibernateException here. What do you think?

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import java.util.function.Function;
+
+public class QueryHelper {
+    private SessionFactory sessionFactory;
+
+    public QueryHelper(SessionFactory sessionFactoryFactory) {
+        this.sessionFactory = sessionFactoryFactory;
+    }
+
+    public <T> T readWithinTx(Function<Session, T> entityManagerConsumer) {
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: Please add unit tests.

Comment (DanielChanger) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: it's not only for read. Please rename this method to runWithinTxReturning, and create another one runWithinTx that returns void

Comment (DanielChanger) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: please make these methods static and QueryHelper constructor private

**Review by DanielChanger — CHANGES_REQUESTED** [review]

**Comment by DanielChanger** [comment]
Write Unit Tests

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+public class QueryHelper {
+
+    private QueryHelper() {
+    }
+
+    public static void runWithinTx(SessionFactory sessionFactory, Consumer<Session> action){
+        runWithinTxReturning(sessionFactory, session -> {
+            action.accept(session);
+            return null;
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java:18]: Just a tip: It doesn't look very good from a reading perspective.As for me, I would rather duplicate the try catch block here.

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        });
+    }
+
+    public static <T> T runWithinTxReturning(SessionFactory sessionFactory, Function<Session, T> sessionConsumer) {
```
Comment (StanislavZabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: Actually we can`t call the second parameter with suffix "consumer" as it is Function, not a consumer

**Review by StanislavZabramnyi — DISMISSED** [review]
👍😉✌️😎

File: src/test/java/com/bobocode/bibernate/integration/QueryHelperIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        runWithinTx(sessionFactory,
+                (session) -> {
+                    Product foundProduct = session.find(Product.class, 1L).orElseThrow();
+                    foundProduct.price(2.0);
+                    session.flush();
+                });
+
+
+        Optional<Product> updatedProduct = runWithinTxReturning(sessionFactory,
+                (session_) -> session_.find(Product.class, 1L));
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/bibernate/integration/QueryHelperIntegrationTest.java]: (nit) _ is typo

File: src/test/resources/persistence.yml [hunk 7]
```diff
@@ -51,3 +51,14 @@ persistenceUnit:
     user: sa
     password: password
   dialect: H2
+logLevel: TRACE
```
Comment (DanielChanger) [inline src/test/resources/persistence.yml]: remove plz

**Review by DanielChanger — CHANGES_REQUESTED** [review]
Add docs plz

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 8]
```diff
@@ -7,18 +7,35 @@
...
+/**
+ * Util class that provides methods for performing logic within session and transaction scope
+ */
 public class QueryHelper {
 
     private QueryHelper() {
     }
 
+    /**
+     * Creates a new session using the provided session factory and performs the passed action within the session scope.
```
Comment (szabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: (nit) would be good to add link to the SessionFactory using {@link SessionFactory}

Comment (szabramnyi) [reply src/main/java/com/bobocode/bibernate/QueryHelper.java]: the same for Session

File: src/main/java/com/bobocode/bibernate/QueryHelper.java [hunk 9]
```diff
@@ -7,18 +7,35 @@
...
+     * Provided logic is run within transaction scope, which will be either committed or rollback (in case exception happens).
```
Comment (szabramnyi) [inline src/main/java/com/bobocode/bibernate/QueryHelper.java]: the same for Transaction

**Review by szabramnyi — DISMISSED** [review]
👍😉

**Review by DanielChanger — DISMISSED** [review]

**Review by StanislavZabramnyi — DISMISSED** [review]
🤗

**Review by DanielChanger — APPROVED** [review]
LGTM, well done

**Pull request merged**